x/auth/src/signing/renderer/value_renderer.rs

```diff
@@ -1,32 +1,49 @@
 //! Trait for formatting all kind of values into `Screen`
 
-use std::error::Error;
-
-use database::Database;
-use gears::types::context::context::Context;
-use proto_messages::cosmos::tx::v1beta1::screen::Screen;
-use store::StoreKey;
+use proto_messages::cosmos::tx::v1beta1::{
+    screen::{Content, Screen},
+    tx_metadata::Metadata,
+};
+use proto_types::Denom;
 
 /// Render primitive type into content for `Screen`.
-/// Use for formatting simple primitive `Copy` types that doesn't require error handling
 pub trait PrimitiveValueRenderer<V> {
-    /// Get string representation of some `V`
-    fn format(value: V) -> String;
+    /// Get string representation of some `V` wrapped in a Content
+    fn format(value: V) -> Content;
+}
+
+pub trait TryPrimitiveValueRenderer<V> {
+    /// Try to get a string representation of some `V` wrapped in a Content
+    fn try_format(value: V) -> Result<Content, Error>;
+}
 
-    /// Try format specific value
-    fn format_try(value: V) -> Result<String, Box<dyn Error>>;
+pub trait TryPrimitiveValueRendererWithMetadata<V> {
+    /// Try to get a string representation of some `V` wrapped in a Content. This method also
+    /// takes a function to get metadata for the denom.
+    fn try_format_with_metadata<F: Fn(&Denom) -> Option<Metadata>>(
+        value: V,
+        get_metadata: &F,
+    ) -> Result<Content, Error>;
 }
 
-/// The notion of "value renderer" is defined in ADR-050.
-pub trait ValueRenderer<SK: StoreKey, DB: Database> {
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("not implemented")]
+    NotImplemented,
```

Comment:
Why we need `NotImplemented` error?

Reply:
Users must implement the ValueRenderer trait on their Messages. This gives them a way out.